### ISSUE TITLE
Research and hopefully make github-access-manager faster

### ISSUE THREAD

**Issue by kucharskim** [ChorusOne/github-access-manager#16]
This step is fast:
```
[1 / ??] Listing organization repositories
```

This step is slow, could it iterate in paralel the repos?
```
[5 / 425] Getting access on ...
```

**Comment by ruuda** [comment]
I looked into it briefly, it’s doing a few requests per repository that we could make in parallel, and also we could request the repositories in parallel, but we’d need to swap out the http logic for something like `aiohttp`, it would be a bit tedious.

**Comment by leo-chorus-one** [comment]
Ok so basically fsck me. Everything is here is GraphQL, perfect, allows to go from 800 requests to a dozen, super fast… except that one small part of getting the teams for a repo. That just doesn’t exist, so back to square one 😭 .

**Comment by ruuda** [comment]
We could maybe do a combination of both? Though maybe it’s not worth the complexity.

**Comment by leo-chorus-one** [comment]
Yes, that’s how it’s going. GraphQL is still worth it for the other queries, and just for this one it’s gonna be threadpool. It’s not so complex but as I have burned 1.5 days on this I’m gonna do something else more useful and then come back to it 😅 .

**Comment by leo-chorus-one** [comment]
@ruuda although, before I sink more time into this lovely hand-rolled python script… why are we not using terraform for this, which would be my most obvious go to for such a problem (the problem the repo solves I mean, not the speed problem) ?

**Comment by ruuda** [comment]
Because Terraform doesn’t actually solve the problem, we looked into it before we wrote this, there are a few plugins that can manage some things on GitHub, but they don’t tell you that we have people in the organization, or as collaborators on repos, who should not have access. I also thought something like it must exist, but if it did, I couldn’t find it, so we wrote it.

**Comment by leo-chorus-one** [comment]
Interesting, ok… that may have been the early days of the provider when https://registry.terraform.io/providers/integrations/github/latest/docs/resources/repository_collaborators (which handles individuals and teams, and is authoritative) did not exist. But maybe I don’t understand the problem you had, or you had other problems.

**Comment by ruuda** [comment]
If I recall correctly, the problem with that one is more of a problem with Terraform’s model of the world. You can define a <del>collaborator</del><ins>organization member</ins>, and if it didn’t exist, it will create it. And if you later remove it from your config, it will still be in the Terraform state, and so it will offer to remove it. But if somebody outside of Terraform just added a <del>collaborator</del><ins>member</ins> in the GitHub settings, then Terraform will not be aware of that.

**Comment by leo-chorus-one** [comment]
No no, that’s exactly what I mean by "it’s authoritative" ! For `github_repository_collaborators` the resource fully manages the list of collaborators and teams for one repo, which is what we want. You also have the non-authoritative versions of the same resource, `github_repository_collaborator` and `github_team_repository`, which act like you describe.

This authoritative/non-authoritative duality comes up often in Terraform, see similar issue @alphadose had: https://github.com/ChorusOne/chorus-infrastructure/pull/15435#issuecomment-2493331274.

**Comment by leo-chorus-one** [comment]
No no, that’s exactly what I mean by "it’s authoritative" ! For github_repository_collaborators the resource fully manages the list of collaborators and teams for one repo, which is what we want. You also have the non-authoritative versions of the same resource, github_repository_collaborator and github_team_repository, which act like you describe.
